### ISSUE TITLE
SSL test scripts fail out of the box on Ubuntu 18.04 and newer

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#1785]
Note that this is not a bug in the Mbed TLS library, it's a bug in the test scripts.

On Debian stretch (currently the stable release of Debian) and Ubuntu 18.04 bionic (currently the stable release of Ubuntu), `ssl-opt.sh` and `compat.sh` fail out of the box. This turns out to be due to two issues related to the `openssl` command line tool.

### `openssl s_server -dhparam`

The option `-dhparam` to `openssl s_server` doesn't work on Debian stretch which has OpenSSL 1.1.0f. It does work on Ubuntu bionic which has OpenSSL 1.1.0g.
```
openssl s_server -www -cert data_files/server5.crt -key data_files/server5.key -accept 17183 -dhparam data_files/dhparams.pem -key data_files/server2.key              -cert data_files/server2.ku-ds.crt
Error with command: "-dhparam data_files/dhparams.pem"
```
I think this is due to OpenSSL commit https://github.com/openssl/openssl/commit/7e1b7485706c2b11091b5fa897fe496a2faa56cc which broke `-dhparam`. It was fixed in https://github.com/openssl/openssl/commit/7e1b7485706c2b11091b5fa897fe496a2faa56cc for OpenSSL 1.1.0g and https://github.com/openssl/openssl/commit/51ac82702dc91cabd3dbf890d8f65b285282c0ce for OpenSSL 1.1.1.

Fortunately the only reason we pass `-dhparam` is because it was needed with older versions of OpenSSL (≤ 1.0.2a): 6195767554da332e9f81e6510b07f7565ff8a538. The following solutions work:

* Keep it simple: abandon support for OpenSSL ≤1.0.2a in `compat.sh` and `ssl-opt.sh`.

    ```
    diff a/tests/compat.sh b/tests/compat.sh
    --- a/tests/compat.sh
    +++ b/tests/compat.sh
    @@ -843,7 +843,7 @@
         esac

         M_SERVER_ARGS="server_port=$PORT server_addr=0.0.0.0 force_version=$MODE arc4=1"
    -    O_SERVER_ARGS="-accept $PORT -cipher NULL,ALL -$MODE -dhparam data_files/dhparams.pem"
    +    O_SERVER_ARGS="-accept $PORT -cipher NULL,ALL -$MODE"
         G_SERVER_ARGS="-p $PORT --http $G_MODE"
         G_SERVER_PRIO="NORMAL:+ARCFOUR-128:+NULL:+MD5:+PSK:+DHE-PSK:+ECDHE-PSK:+RSA-PSK:-VERS-TLS-ALL:$G_PRIO_MODE"

    diff a/tests/ssl-opt.sh b/tests/ssl-opt.sh
    --- a/tests/ssl-opt.sh
    +++ b/tests/ssl-opt.sh
    @@ -693,7 +693,7 @@ SRV_DELAY_SECONDS=0
     P_SRV="$P_SRV server_addr=127.0.0.1 server_port=$SRV_PORT"
     P_CLI="$P_CLI server_addr=127.0.0.1 server_port=+SRV_PORT"
     P_PXY="$P_PXY server_addr=127.0.0.1 server_port=$SRV_PORT listen_addr=127.0.0.1 listen_port=$PXY_PORT ${SEED:+"seed=$SEED"}"
    -O_SRV="$O_SRV -accept $SRV_PORT -dhparam data_files/dhparams.pem"
    +O_SRV="$O_SRV -accept $SRV_PORT"
     O_CLI="$O_CLI -connect localhost:+SRV_PORT"
     G_SRV="$G_SRV -p $SRV_PORT"
     G_CLI="$G_CLI -p +SRV_PORT localhost"
    ```

* Pass `-dhparam` only to older OpenSSL. I tested that in https://github.com/gilles-peskine-arm/mbedtls/tree/openssl-dhparam-only_if_needed and https://github.com/gilles-peskine-arm/mbedtls/tree/openssl-dhparam-only_if_needed-2.11.

Note that our reference platform Ubuntu 16.04 xenial is recent enough not to need `-dhparam`, and old enough that `-dhparam` wasn't broken yet. Our upcoming reference platform Ubuntu 18.04 is recent enough to have a repaired `-dhparam`.

### Obsolete ciphersuites removed

OpenSSL 1.1.0f in Debian stretch and OpenSSL 1.1.0g in Ubuntu bionic no longer support 3DES ciphersuites and non-ephemeral (EC)DH ciphersuites, and no longer support SSLv3. I haven't verified whether this is a change in the Debian configuration or an OpenSSL upstream change.

After removing `-dhparam` from the `openssl s_server` calls, the following call passes on Debian stretch:
```
tests/compat.sh -p OpenSSL -e '\(^\|-\)\(NULL\|ECDH\|DES\|3DES\|RC4\|ARCFOUR\|ARIA\)[0-9]*-' -m 'tls1 tls1_1 tls1_2 dtls1 dtls1_2 '
```

I think we should make `compat.sh` filter on the ciphersuites that the available `openssl` program supports. We can run `openssl ciphers ALL` to list them.

Note that this likely means that on the CI, we'll need to test with several different builds of OpenSSL — a recent one for recent ciphers like ARIA, and an older version for legacy ciphers like 3DES. Ping @mazimkhan

**Comment by gilles-peskine-arm** [comment]
Ah, right! I wasn't thinking of the legacy OpenSSL version, thanks. OpenSSL 1.0.1j requires `-dhparam` so we need to keep adding it when needed in the test scripts.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-2523

**Comment by gilles-peskine-arm** [comment]
@RonEld Why did you close this? Is it resolved? I haven't tried recently.

**Comment by gilles-peskine-arm** [comment]
I've checked and `compat.sh` still fails out of the box on Debian stretch, so I'm reopening, otherwise we'll keep running against this and rediscovering it. I'm changing the label to "enhancement" since we don't advertise support for any specific platform.

**Comment by gilles-peskine-arm** [comment]
The `-dhparam` issue was fixed in #4429.

We still need to do something about static ECDH. Either add it to the default exclusion list, or add some form of openssl version detection.

**Comment by gilles-peskine-arm** [comment]
After a short discussion on Slack, Manuel both think that `compat.sh` should detect the version of OpenSSL, and skip cipher suites that it doesn't support. So:

* In `setup_arguments`, set a variable that indicates whether OpenSSL supports static ECDH.
* In `run_client`, before checking `SKIP_NEXT`, add a check: if the cipher suite is static ECDH and the program is OpenSSL, decide to skip.
